### PR TITLE
[bugfix] fix query_frontend init fail error

### PR DESCRIPTION
**Pull request by liguozhong**
**What this PR does / why we need it**:
My loki cluster is distributed without a compactor, but compactorAddress must require me to fill in a compactor url.
Using the latest version of the master branch, the query frontend of loki cluster cannot be started and will encounter this error log.
```go
query filtering for deletes requires 'compactor_address' to be configured
```

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:
my loki cluster Topology.
image

**Checklist**
- [ ] Documentation added
- [ ] Tests updated
- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Comment by jeschkies** [comment]
Please merge main to fix the failing `helm-ci` job.

**Comment by liguozhong** [comment]
> Please merge main to fix the failing `helm-ci` job.

done，thanks

**Comment by ashwanthgoli** [comment]
Returning the address of localhost is only valid for single binary or ssd deployments. For multi-store index (as it does not support delete requests), query-frontend need not make calls to compactor. So a better fix for this issue would be to pass down a nil `cacheGenNumLoader` to `NewTripperware` func for indexes that don't support deletions
https://github.com/grafana/loki/blob/e39ea21e3cacda92661d1d77770c7d1486ec4124/pkg/querier/queryrange/roundtrip.go#L41-L46

It would be nice if you could also refactor `cacheGenClient()` to return `*GenNumberLoader` instead. It can include the logic to return nil based on the index type (for TSDB and multi-store)
https://github.com/grafana/loki/blob/e39ea21e3cacda92661d1d77770c7d1486ec4124/pkg/loki/modules.go#L1067-L1074

Thanks for taking time out and submitting the PR :)

**Comment by liguozhong** [comment]
> Returning the address of localhost is only valid for single binary or ssd deployments. For multi-store index (as it does not support delete requests), query-frontend need not make calls to compactor. So a better fix for this issue would be to pass down a nil `cacheGenNumLoader` to `NewTripperware` func for indexes that don't support deletions
> 
> https://github.com/grafana/loki/blob/e39ea21e3cacda92661d1d77770c7d1486ec4124/pkg/querier/queryrange/roundtrip.go#L41-L46
> 
> It would be nice if you could also refactor `cacheGenClient()` to return `*GenNumberLoader` instead. It can include the logic to return nil based on the index type (for TSDB and multi-store)
> 
> https://github.com/grafana/loki/blob/e39ea21e3cacda92661d1d77770c7d1486ec4124/pkg/loki/modules.go#L1067-L1074
> 
> Thanks for taking time out and submitting the PR :)

hi, thank you for such a detailed review suggestion.
I resubmitted the code once as you suggested, if I understand correctly

File: pkg/loki/modules.go [hunk 1]
```diff
@@ -654,13 +653,26 @@ func (t *Loki) initQueryFrontendTripperware() (_ services.Service, err error) {
 	return services.NewIdleService(nil, nil), nil
 }
 
-func (t *Loki) cacheGenClient() (generationnumber.CacheGenClient, error) {
+func (t *Loki) cacheGenClient() (*generationnumber.GenNumberLoader, error) {
+	if config.UsingTSDB(t.Cfg.SchemaConfig.Configs) {
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (liguozhong) [reply pkg/loki/modules.go]: hi , according to your suggestion, I submitted another commit, canceling the code snippets such as ' return nil,nil ' Fragments like this are really bad. 
I operator the loki cluster full-time, this is my job, your suggestions are very helpful to me, thank you for your patience to review and guide

File: pkg/loki/modules.go [hunk 2]
```diff
@@ -654,13 +653,26 @@ func (t *Loki) initQueryFrontendTripperware() (_ services.Service, err error) {
 	return services.NewIdleService(nil, nil), nil
 }
 
-func (t *Loki) cacheGenClient() (generationnumber.CacheGenClient, error) {
+func (t *Loki) cacheGenClient() (*generationnumber.GenNumberLoader, error) {
```
Comment (MasslessParticle) [inline pkg/loki/modules.go]: This should probably just return the interface type `queryrangebase.CacheGenNumberLoader`

Comment (MasslessParticle) [inline pkg/loki/modules.go]: Let's also rename the function to `cacheGenNumberLoader`

**Review by MasslessParticle — APPROVED** [review]

**Pull request merged**